crates/wasmlink/runtime/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+        if self.next == self.storage.len() {
+            self.storage.push(Entry::Empty {
+                next: self.next + 1,
+            });
+        }
+        let ret = self.next as u32;
+        let entry = Entry::Full(item);
+        self.next = match mem::replace(&mut self.storage[self.next], entry) {
+            Entry::Empty { next } => next,
+            _ => unreachable!(),
+        };
+        return ret;
+    }
+
+    pub fn get(&self, idx: u32) -> Option<&T> {
+        match self.storage.get(idx as usize)? {
+            Entry::Full(b) => Some(b),
+            Entry::Empty { .. } => None,
+        }
+    }
+
+    pub fn get_mut(&mut self, idx: u32) -> Option<&mut T> {
+        match self.storage.get_mut(idx as usize)? {
+            Entry::Full(b) => Some(b),
+            Entry::Empty { .. } => None,
+        }
+    }
+
+    pub fn remove(&mut self, idx: u32) -> Option<T> {
+        let slot = self.storage.get_mut(idx as usize)?;
+        match mem::replace(slot, Entry::Empty { next: self.next }) {
+            Entry::Full(b) => {
+                self.next = idx as usize;
+                Some(b)
+            }
+            Entry::Empty { next } => {
+                *slot = Entry::Empty { next };
+                None
+            }
+        }
+    }
+}
+
+impl<T> Default for Slab<T> {
+    fn default() -> Slab<T> {
+        Slab {
+            storage: Vec::new(),
+            next: 0,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct IndexSlab {
+    slab: Slab<ResourceIndex>,
+}
+
+impl IndexSlab {
+    pub fn insert(&mut self, resource: ResourceIndex) -> u32 {
+        self.slab.insert(resource)
+    }
+
+    pub fn get(&self, slab_idx: u32) -> Option<ResourceIndex> {
+        self.slab.get(slab_idx).cloned()
+    }
+
+    pub fn remove(&mut self, slab_idx: u32) -> Option<ResourceIndex> {
+        self.slab.remove(slab_idx)
+    }
+}
+
+#[derive(Default)]
+pub struct ResourceSlab {
+    slab: Slab<Resource>,
+}
+
+#[derive(Debug)]
+struct Resource {
+    wasm: u32,
+    references: u32,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct ResourceIndex(u32);
+
+impl ResourceSlab {
+    pub fn insert(&mut self, wasm: u32) -> ResourceIndex {
+        ResourceIndex(self.slab.insert(Resource {
+            wasm,
+            references: 1,
+        }))
+    }
+
+    pub fn get(&self, idx: ResourceIndex) -> u32 {
+        self.slab.get(idx.0).unwrap().wasm
+    }
+
+    pub fn clone(&mut self, idx: ResourceIndex) {
+        let resource = self.slab.get_mut(idx.0).unwrap();
+        resource.references = resource.references.checked_add(1).unwrap();
+    }
+
+    pub fn drop(&mut self, idx: ResourceIndex) -> Option<u32> {
+        let resource = self.slab.get_mut(idx.0).unwrap();
+        assert!(resource.references > 0);
+        resource.references -= 1;
+        if resource.references != 0 {
+            return None;
+        }
+        let resource = self.slab.remove(idx.0).unwrap();
```

Comment:
There's a good number of `unwrap()` in this module which is used to abort the process, which while semantically fine makes it hard to distinguish an `unwrap()` like this where it's a bug in this module if this trips. Could a separate function be added which is specifically used for `unreachable`-on-bad-inputs? We could presumably give it a name that's indicative of "this is the consuming module's fault" and the implementation of the function could be to just call [`unreachable`](https://doc.rust-lang.org/core/arch/wasm32/fn.unreachable.html) (since presumably this module doesn't import WASI)

Reply:
I think that makes as lot of sense to at least indicate somehow that the problem was coming from the user's code even through it was the wasmlink runtime that trapped.

Presumably without WASI, the function name would need to suffice to communicate this, so perhaps we should keep the name section for the runtime after all.

Reply:
Yeah that's a good point, and I was thinking that the name section would be the way to communicate this, yeah.

I'm also sort of naively assuming that we *can't* use WASI here since the linker may produce binaries that don't want to use wasi at all, but we could possibly have a flag where if wasi is used (or detected already) then we could have an alternate version here which prints a more descriptive error. Anyway lots of future possibilities.

Reply:
I've added a function named `invalid_handle_value` that will show up in the backtrace when the user passes an invalid index.